### PR TITLE
Train on GPUs with PyTorch

### PR DESCRIPTION
**Pull request by rohskopf**
This PR gets GPU training working in PyTorch. Speedups are only noticeable when using large models or batches, which is typical of GPU acceleration for training. 

**Example benchmarks:**
Ta example with layer sizes of `1000x800x600x1`, `batch_size=4`: 
- GPU (GTX 1080): ~0.20 s/epoch
- CPU: ~2.0 s/epoch

Ta example with layer sizes of `10x8x6x1`, `batch_size=4`: 
- GPU (GTX 1080): ~0.20 s/epoch
- CPU: ~0.20 s/epoch

This suggests that we can benefit from parallelization that is more suitable for large datasets and small models, e.g. nicely documented here: https://pytorch.org/docs/stable/distributed.html

One option to speed up training is to parallelize over batches on separate processes, and average gradients on those processes back to the head process for updating the weights. This could be done using GPU or CPU nodes on a cluster, and should be embarrassingly parallel.

**Challenges:**
1. We get great results with small batch sizes (<= 4), so we can't benefit much from data parallelism there.
2. We have small models (<100k params), so we can't benefit much from GPU speed up there.
3. To take advantage of GPUs and data parallelism, we need larger batch sizes, need to find settings that learn well there.

We also have poor GPU utilization when training; the `nvidia-smi` command in Linux gives a utilization of < 10% when using small models and only ~30-40% for larger models. Utilization > 90% is desirable. This suggests that there are some CPU bottlenecks.

**CPU bottlenecks:**
1. In `dataloaders.py` where we collate large amounts of data needed for batch training.
2. In `solvers/pytorch.py` and `lib/neural_networks/pytorch.py` where we send data from CPU to GPU (which is a slow process) in a loop. It would be better to calculate bispectrum components on GPU directly, or somehow transfer descriptors to the GPU before training occurs.

Nonetheless this PR has GPU training working, and it can be improved upon.

### CODE REVIEW

**Review by mitwood — APPROVED** [review]
Nice additions allowing for GPU acceleration of NN fitting. Work in progress as outline in PR comments from @rohskopf.
